src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -109,6 +117,96 @@ public void write(String str, int off, int len) throws IOException {
         }  
     }
 
+    private class UnfinishedLine {
+        final long threadId;
+        int curCapacity = 512;
+        byte[] buffer = new byte[curCapacity];
+        int curEnd = 0;
+        int lastNewLine = 0;
+
+        public UnfinishedLine() {
+            this.threadId = Thread.currentThread().getId();
+        }
+
+        /**
+         * Adding input combines previously unfinished sentences with possible
+         * new sentences. A number of cases come together here that otherwise
+         * should be diligently separated. 
+         * 
+         *   - An unfinished line can already exist or not
+         *   - The new input can contain newlines or not
+         *   - The new input can end with a newline character or not
+         * 
+         * By concatenating the previous unfinished line with the new input
+         * all we have to do now is figure out where the last newline character is.
+         * 
+         */
+        private void add(byte[] newInput, int offset, int len) {
+            // first ensure capacity of the array
+            if (curEnd + len >= curCapacity) {
+                var oldCapacity = curCapacity;
+                curCapacity *= 2; // this should not happen to often. we're talking a few lines of text.
+                byte[] tmp = new byte[curCapacity];
+                System.arraycopy(buffer, 0, tmp, 0, oldCapacity);
+                buffer = tmp;
+            }
+
+            System.arraycopy(newInput, offset, buffer, curEnd, len);
+            curEnd += len;
+            lastNewLine = startOfLastLine();
```

Comment:
the interaction between `startOfLastLine` and `flushToLastLine` is a bit finicky. also maybe a bit wasteful (copying to the array (potentiially growing it) when it might not be needed?

Also, `startOfLastLine` can be helped by the fact that it only has to scan the just in buffer.

I think I would do: 
1. scan for last newline in `newInput`, 
2. if it contains it:
   1.  flush the current `buffer` to the stream
   2. flush everything in `newInput` to the newline
3. copy everything from the `newInput` past the last newline into the `buffer` array.

this also means we don't have to keep that `lastNewLine` variable around?

Reply:
I'll think  about it. The current implementation deals with all possible situations of newlines being present or not, in previous input or in new input. I think your way can also work, but I need some new invariants to make sure it is correct in all cases. I think the new invariant would be that the buffered input does not contain newlines at all. I'll ponder this for a bit.

Reply:
There is also a fast path possible for input that ends in a '\n' when the current buffer is empty.